### DISCUSSION TITLE
How to get diff of two images?

### DISCUSSION THREAD

**Discussion by rabbitLove520** [dlemstra/Magick.NET#1293] [Help]
Two images, same size, I want to calculate the rgb value difference of every pixel, any built-in method?

**Comment by dlemstra** [comment]
You could use the `Compare` method of `MagickImage` for this?

**Reply by rabbitLove520** [reply]
`Compare` is not what I want, I want the method like this
```

```
private static MagickImage imageDiff(IPixelCollection<ushort> pc1, IPixelCollection<ushort> pc2, int width, int height)
{
    var result = new MagickImage(MagickColor.FromRgb(0, 0, 0), width, height);

    for (int x = 0; x < width; x++)
    {
        for (int y = 0; y < height; y++)
        {
            var p1 = pc1[x, y];
            var p2 = pc2[x, y];

            var arr1 = p1.ToArray();
            var arr2 = p2.ToArray();

            var span = new ushort[4];
            for (int i = 0; i < span.Length; i++)
            {
                var tmp = arr1[i] - arr2[i];
                if (tmp <= 0)
                {
                    tmp = 0;
                }
                span[i] = Convert.ToUInt16(tmp);
            }
            
            result.GetPixels().SetPixel(x, y, span);
        }
    }
    return result;
}
```
```

**Comment by rabbitLove520** [comment]
this method implemented by myself is ok, however, it's slow...
```
private static MagickImage imageDiff(IPixelCollection<ushort> pc1, IPixelCollection<ushort> pc2, int width, int height)
{
    var result = new MagickImage(MagickColor.FromRgb(0, 0, 0), width, height);

    for (int x = 0; x < width; x++)
    {
        for (int y = 0; y < height; y++)
        {
            var p1 = pc1[x, y];
            var p2 = pc2[x, y];

            var arr1 = p1.ToArray();
            var arr2 = p2.ToArray();

            var span = new ushort[4];
            for (int i = 0; i < span.Length; i++)
            {
                var tmp = arr1[i] - arr2[i];
                if (tmp <= 0)
                {
                    tmp = 0;
                }
                span[i] = Convert.ToUInt16(tmp);
            }
            
            result.GetPixels().SetPixel(x, y, span);
        }
    }
    return result;
}
```

**Reply by dlemstra** [reply]
You should get an improvement when you do `result.GetPixels()` only once and `result.GetPixelsUnsafe()` should be even faster.

But it should also be possible to use the `-fx` operator on multiple images (https://imagemagick.org/script/fx.php). That feature is not yet available in this library and I think we will need to and an extra overload to `MagickImage.Fx` to make this possible.

**Reply by dlemstra** [reply]
The `Fx` will become available for the `MagickImageCollection` in the next release.
